src/NotepadNext/docks/FolderAsWorkspaceDock.cpp

```diff
@@ -65,3 +74,95 @@ QString FolderAsWorkspaceDock::rootPath() const
 {
     return model->rootPath();
 }
+
+void FolderAsWorkspaceDock::onCustomContextMenu(const QPoint &point)
+{
+    QModelIndex index = ui->treeView->indexAt(point);
+    if (!index.isValid()) {
+        lastSelectedItem = model->index(rootPath());
+        ui->menuEmpty->exec(ui->treeView->viewport()->mapToGlobal(point));
+        return;
+    }
+    lastSelectedItem = index;
+    if (model->isDir(index)) {
+        ui->menuDirectory->exec(ui->treeView->viewport()->mapToGlobal(point));
+    }
+    else {
+        ui->menuFile->exec(ui->treeView->viewport()->mapToGlobal(point));
+    }
+}
+
+void FolderAsWorkspaceDock::on_actionSaveHere_triggered()
+{
+    QDir parentDir(model->filePath(lastSelectedItem));
+    auto doc = window->currentEditor();
+    QString dstName(parentDir.absoluteFilePath(doc->getName()));
+
+    if (doc->saveAs(dstName) != QFileDevice::NoError)
```

Comment:
In most cases I prefer the "happy path" to be the first part (e.g. switching this to `==`). Just helps with readability since you have to skip over the failure case and look at the `else` clause to see the natural flow.

Also, this is a bit of a code 'smell' that the FolderAsWorkspace is messing with the lower editor API. Might be possible to use something like `window->saveFileAs(...)` so that incase there are errors it will be handled the same way as a normal file would if it failed.

Reply:
I separate UI questions inside MainWindow and use them here instead.